servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/JavaNetSoTimeoutHttpConnectionFilter.java

```diff
@@ -167,24 +174,99 @@ public Single<StreamingHttpResponse> request(final StreamingHttpRequest request)
-                                            .initCause(t))))
+                            .<StreamingHttpResponse>liftSync(subscriber ->
+                                    new RequestTimeoutSubscriber(subscriber,
+                                            SourceAdapters.fromSource(requestProcessor), timeout, timeoutExecutor))
                             .shareContextOnSubscribe();
                 });
             }
         };
     }
 
+    // package private for testing purposes
+    static final class RequestTimeoutSubscriber implements Subscriber<StreamingHttpResponse> {
+
+        private static final AtomicIntegerFieldUpdater<RequestTimeoutSubscriber> onceUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(RequestTimeoutSubscriber.class, "once");
+
+        private final DelayedCancellable requestCancellable = new DelayedCancellable();
+        private final Cancellable timeoutCancellable;
+        private final Subscriber<? super StreamingHttpResponse> delegate;
+
+        private final Duration timeout;
+        private final Executor timeoutExecutor;
+        @SuppressWarnings("unused")
+        private volatile int once;
+
+        RequestTimeoutSubscriber(Subscriber<? super StreamingHttpResponse> delegate, Completable requestComplete,
+                                 Duration timeout, Executor timeoutExecutor) {
+            this.delegate = delegate;
+            this.timeout = timeout;
+            this.timeoutExecutor = timeoutExecutor;
+            timeoutCancellable = requestComplete.concat(Completable.never()
+                    .timeout(timeout, timeoutExecutor)).beforeOnError(this::handleInterruptions).subscribe();
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            requestCancellable.delayedCancellable(cancellable);
+            try {
+                delegate.onSubscribe(() -> {
+                    once();
+                    timeoutCancellable.cancel();
+                    requestCancellable.cancel();
+                });
+            } catch (Throwable cause) {
+                handleExceptionFromOnSubscribe(this, cause);
+                cancellable.cancel();
+            }
+        }
+
+        private void handleInterruptions(Throwable t) {
+            if (once()) {
+                requestCancellable.cancel();
+                Throwable result = t;
+                // We can get a SocketTimeoutException waiting for a 100 Continue response.
+                if (t instanceof TimeoutException) {
+                    result = newStacklessSocketTimeoutException("Read timed out after " + timeout.toMillis() +
+                            "ms waiting for response meta-data").initCause(t);
+                }
+                delegate.onError(result);
+            }
+        }
+
+        @Override
+        public void onSuccess(@Nullable StreamingHttpResponse result) {
+            if (once()) {
+                timeoutCancellable.cancel();
```

Comment:
For this one there is less risk, because the only way for users code to be in this path is if users supply a custom `Executor` that can throw on `cancel()`. Don't have a strong opinion here. If you decide to try-catch this one, please do the same here and in `onError`.